x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -173,69 +174,169 @@ private AuthorizationInfo getAuthorizationInfoFromContext() {
                 // this never goes async so no need to wrap the listener
                 authorizeSystemUser(authentication, action, auditId, unwrappedRequest, listener);
             } else {
-                final String finalAuditId = auditId;
-                final RequestInfo requestInfo = new RequestInfo(authentication, unwrappedRequest, action);
                 final ActionListener<AuthorizationInfo> authzInfoListener = wrapPreservingContext(ActionListener.wrap(
                         authorizationInfo -> {
                             threadContext.putTransient(AUTHORIZATION_INFO_KEY, authorizationInfo);
-                            maybeAuthorizeRunAs(requestInfo, finalAuditId, authorizationInfo, listener);
+                            maybeAuthorizeRunAs(requestInfo, auditId, authorizationInfo, listener);
                         }, listener::onFailure), threadContext);
-                getAuthorizationEngine(authentication).resolveAuthorizationInfo(requestInfo, authzInfoListener);
+                engine.resolveAuthorizationInfo(requestInfo, authzInfoListener);
+            }
+        }
+    }
+
+    /**
+     * @return {@code true} if this action is a child action of an already authorized action that is in the thread context
+     * <strong>and</strong> we can infer that this action is authorized from the parent action.
+     */
+    private boolean shouldAuthorizeAsChildAction(AuthorizationEngine engine, RequestInfo requestInfo) {
+        if (SystemUser.is(requestInfo.getAuthentication().getUser())) {
+            // System user requests are already handled efficiently and never use an AuthorizationEngine
+            return false;
+        }
+        if (isIndexAction(requestInfo.getAction()) == false) {
+            // We depend on there being an IndicesAccessControl, so only consult the engine for index actions
+            return false;
+        }
+
+        final String parentAction = threadContext.getTransient(ORIGINATING_ACTION_KEY);
+        if (Strings.isNullOrEmpty(parentAction)) {
+            // No parent action
+            return false;
+        }
+
+        final IndicesAccessControl parentAccessControl = threadContext.getTransient(AuthorizationServiceField.INDICES_PERMISSIONS_KEY);
+        if (parentAccessControl == null) {
+            // This is almost certainly an error, but leave it to be handled later
+            return false;
+        }
+
+        if (engine.isChildActionAuthorizedByParent(requestInfo, parentAction, parentAccessControl)) {
+            logger.trace(
+                "Automatically authorizing child action [{}] (child of [{}]) with request [{}]",
+                requestInfo.getAction(),
+                parentAction,
+                requestInfo.getRequest()
+            );
+            return true;
+        }
+        return false;
+    }
+
+    private void authorizeChildAction(AuthorizationEngine engine, RequestInfo requestInfo, String auditId, ActionListener<Void> listener) {
```

Comment:
I think we should add a comment on the fact that run-as does not need to be checked for child actions and hence auditing for run-as of child actions is skipped as well.